src/Microsoft.Deployment.DotNet.Releases/src/ReleaseVersion.cs

```diff
@@ -568,5 +547,376 @@ internal static int CompareNumericIdentifier(string a, string b)
+
+                    prerelease = input.Substring(prereleaseSeparatorIndex + 1);
+                    versionCore = input.Substring(0, prereleaseSeparatorIndex);
+                }
+            }
+            else
+            {
+                if (prereleaseSeparatorIndex == -1 || prereleaseSeparatorIndex > buildSeparatorIndex)
+                {
+                    // <valid semver> ::= <version core> "+" <build>
+                    if (buildSeparatorIndex + 1 > input.Length)
+                    {
+                        if (throwOnFailure)
+                        {
+                            throw new FormatException(ReleasesResources.BuildMetadataCannotBeEmpty);
+                        }
+
+                        return false;
+                    }
+
+                    build = input.Substring(buildSeparatorIndex + 1);
+                    versionCore = input.Substring(0, buildSeparatorIndex);
+                }
+                else
+                {
+                    // <valid semver> ::= <version core> "-" <pre-release> "+" <build>
+                    if (prereleaseSeparatorIndex + 1 < buildSeparatorIndex)
+                    {
+                        if (buildSeparatorIndex + 1 > input.Length)
+                        {
+                            if (throwOnFailure)
+                            {
+                                throw new FormatException(ReleasesResources.BuildMetadataCannotBeEmpty);
+                            }
+
+                            return false;
+                        }
+
+                        build = input.Substring(buildSeparatorIndex + 1);
+                        prerelease = input.Substring(prereleaseSeparatorIndex + 1, buildSeparatorIndex - prereleaseSeparatorIndex - 1);
+                        versionCore = input.Substring(0, prereleaseSeparatorIndex);
+                    }
+                    else
+                    {
+                        if (throwOnFailure)
+                        {
+                            throw new FormatException(ReleasesResources.PrereleaseCannotBeEmpty);
+                        }
+
+                        return false;
+                    }
+                }
+            }
+
+            string[] versionParts = versionCore.Split('.');
+
+            if (versionParts.Length != 3)
+            {
+                // <version core> ::= <major> "." <minor> "." <patch>
+                if (throwOnFailure)
+                {
+                    throw new FormatException(string.Format(ReleasesResources.InvalidVersion, versionCore));
+                }
+
+                return false;
+            }
+
+            int[] versions = new int[3] { 0, 0, 0 };
+
+            for (int i = 0; i < 3; i++)
+            {
+                if (!TryParseCoreVersionPart(versionParts[i], throwOnFailure, out versions[i]))
+                {
+                    return false;
+                }
+            }
+
+            if (!IsValidPreleae(prerelease))
```

Comment:
It looks like both places that call `IsValidPrerelease` throw the same exception.  Have you considered refactoring the method to throw - e.g. IsValidPrerelease(prerelease, throwOnFailure)?  It would simplify the calling logic in both places and better encapsulate the exception logic to ensure consistency.

Reply:
Fixed. I did the same for `IsValidBuildMetadata`